core/src/subject/SubjectEntity.ts

```diff
@@ -49,75 +231,192 @@ export class SubjectEntity {
+      ${subQueries.join(", ")}
+    `;
+
+    const result = await this.#perspective.infer(fullQuery);
+    if (result?.[0]) {
+      const { Properties, Collections, Timestamp, Author } = result?.[0];
+      const values = [...Properties, ...Collections, ["timestamp", Timestamp], ["author", Author]];
+      await SubjectEntity.assignValuesToInstance(this.#perspective, this, values);
+    }
+
+    return this;
+  }
+
+  // static async findOne(perspective: PerspectiveProxy, query?: any) {
+  //   let prologQuery = await this.getPropertiesQuery(perspective);
+  //   const result = (await perspective.infer(prologQuery))[0];
+  // }
+
+  /**
+   * Gets all instances of the subject entity in the perspective that match the query params .
+   * @param perspective - The perspective that the subject entities belongs to.
+   * @param query - The query object used to define search contraints.
+   *
+   * @returns An array of all subject entity matches.
+   *
+   */
+  static async findAll(perspective: PerspectiveProxy, query?: Query) {
+    // todo:
+    // + order
+    // + limit
+    // + offset
+    // + include
+
+    const subQueries = [
+      buildSourceQuery(query?.source),
+      buildAuthorAndTimestampQuery(query?.properties),
+      buildPropertiesQuery(query?.properties),
+      buildCollectionsQuery(query?.collections),
+      buildWhereQuery(query?.where),
+    ];
+
+    const fullQuery = `
+      findall([Base, Properties, Collections, Timestamp, Author], (
+        subject_class("${await this.getClassName(perspective)}", SubjectClass),
+        instance(SubjectClass, Base),
+        ${subQueries.filter((q) => q).join(", ")}
+      ), AllInstances)
+    `;
+
+    if (query?.where) console.log("whereQuery!!!", buildWhereQuery(query?.where));
+
+    const result = await perspective.infer(fullQuery);
+    if (!result?.[0]?.AllInstances) return [];
+
+    // Map results to instances
+    const requestedAttribtes = [...(query?.properties || []), ...(query?.collections || [])];
+    const allInstances = await Promise.all(
+      result[0].AllInstances.map(async ([Base, Properties, Collections, Timestamp, Author]) => {
+        const instance = new this(perspective, Base) as any;
+        // Remove unrequested attributes from instance
+        if (requestedAttribtes.length) {
+          Object.keys(instance).forEach((key) => {
+            if (!requestedAttribtes.includes(key)) delete instance[key];
+          });
+        }
+        // Collect values to assign to instance
+        const values = [...Properties, ...Collections];
+        if (!query?.properties || query.properties.includes("timestamp")) values.push(["timestamp", Timestamp]);
+        if (!query?.properties || query.properties.includes("author")) values.push(["author", Author]);
+        await SubjectEntity.assignValuesToInstance(perspective, instance, values);
+
+        return instance;
+      })
+    );
+
+    return allInstances;
```

Comment:
👏 👏 👏 
I guess next step could be to extract the prolog result parsing into its own function, well basically all 3 steps (query creation, infer, result parsing), so that we can have a version that only runs once, and one that subscribes to the query which we want to use in the hooks.